src/test/java/com/google/devtools/build/lib/remote/SimpleBlobStoreFactoryTest.java

```diff
@@ -0,0 +1,190 @@
+package com.google.devtools.build.lib.remote;
+
+import com.google.devtools.build.lib.remote.blobstore.CombinedDiskHttpBlobStore;
+import com.google.devtools.build.lib.remote.blobstore.OnDiskBlobStore;
+import com.google.devtools.build.lib.remote.blobstore.SimpleBlobStore;
+import com.google.devtools.build.lib.remote.blobstore.http.HttpBlobStore;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import com.google.devtools.common.options.Options;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.io.IOException;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+
+@RunWith(JUnit4.class)
+public class SimpleBlobStoreFactoryTest {
+
+  @Rule
+  public ExpectedException expectedEx = ExpectedException.none();
+
+  @Mock
+  private Path workingDirectory;
+
+  @Mock
+  private Path relativeDirectory;
+
+  @Before
+  public final void setUp() {
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test
+  public void createCombinedCache() throws IOException {
+    RemoteOptions options = Options.getDefaults(RemoteOptions.class);
+    options.remoteHttpCache = "http://doesnotesist.com";
+    options.diskCache = PathFragment.create("/etc/something/cache/here");
+
+    when(workingDirectory.getRelative(options.diskCache)).thenReturn(relativeDirectory);
+    when(relativeDirectory.exists()).thenReturn(true);
+
+    SimpleBlobStore blobStore = SimpleBlobStoreFactory.create(options, /* creds */ null, workingDirectory);
+
+    assertThat(blobStore).isNotNull();
+    assertThat(blobStore).isInstanceOf(CombinedDiskHttpBlobStore.class);
+  }
+
+  @Test
+  public void createCombinedCache_whenWorkingDirectoryDoesNotExist() throws IOException {
+    RemoteOptions options = Options.getDefaults(RemoteOptions.class);
+    options.remoteHttpCache = "http://doesnotesist.com";
+    options.diskCache = PathFragment.create("/etc/something/cache/here");
+
+    when(workingDirectory.getRelative(options.diskCache)).thenReturn(relativeDirectory);
+    when(relativeDirectory.exists()).thenReturn(false);
+
+    SimpleBlobStore blobStore = SimpleBlobStoreFactory.create(options, /* creds */ null, workingDirectory);
+
+    assertThat(blobStore).isNotNull();
+    assertThat(blobStore).isInstanceOf(CombinedDiskHttpBlobStore.class);
+    verify(relativeDirectory, times(1)).createDirectoryAndParents();
+  }
+
+  @Test
+  public void createCombinedCache_whenWorkingDirectoryNull_throwsNPE() throws IOException {
```

Comment:
Would this ever happen when using Bazel? If so throw a checked exception?

Reply:
so, theoretically it is possible, because argument is Nullable. I assume, it has been made nullable because of this line: `blobStore = SimpleBlobStoreFactory.create(remoteOptions, null, null);` in `RemoteWorker` for Http remote cache creation. In other cases - workspace is taken from the environment, so if it is null, then smth trully bad is happening to bazel, maybe NPE is not that bad for such an occasion.

Anyways, I would suggest to keep it as it is for now and review this again in one on the next PRs - I plan to refactor the usage of `SimpleBlobFactory` from the `RemoteWorker` - some logic clearly belongs to the Factory but was not moved there. Just didn't want to make this change too big.

Reply:
I think the proper thing to do would be to add a `workingDirectory != null` check so that it doesn't throw a NPE anymore. This behavior is clearly a bug.

> Just didn't want to make this change too big.

I understand, but this is a one line change :)

Reply:
As discussed offline, will keep NPE for this change and will change this behavior properly with the next PR (requires changes in a few other places, therefore PR would be too big).